Add infrastructure for tracking fusion performance

yiqian1 opened the pull request:
Add tuning fusion test cases in weekly CI.
Testing and reporting of fusion performance are currently disabled until a tuning DB is availabe. Currently the performance report only shows 'TFLops'. May include other performance metrics later.

Tuning (with verification) of fusion/e2e/resnet50 and xmir/e2e/bert-torch-tosa have passed in [this build](http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-1029/4/pipeline).

yiqian1 committed:
Add infrastructure for tracking fusion performance

yiqian1 committed:
Add infrastructure for tracking fusion performance

yiqian1 committed:
Disable fusion performance tests until we have a tuning DB

yiqian1 committed:
Disable fusion performance tests until we have a tuning DB

yiqian1 committed:
Typo

yiqian1 committed:
Typo

yiqian1 committed:
Disable invalid test cases

yiqian1 committed:
Disable invalid test cases

yiqian1 committed:
Restore variable definitions

yiqian1 committed:
Restore variable definitions

giuseros committed:
Use an intermediate buffer to transpose data from global memory to LDS (#1023)

* Use an intermediate buffer to transpose data from global memory to LDS

* bug fixing

* Addressing review feedbacks

---------

Co-authored-by: Giuseppe Rossini <grossini@amd.com>

krzysz00 commented on mlir/tools/rocmlir-tuning-driver/rocmlir-tuning-driver.cpp:
```diff
@@ -336,9 +336,26 @@ int main(int argc, char **argv) {
-  if (failed(runTuningLoop(*source))) {
+
+  ModuleOp module;
+  WalkResult findModule = source->walk([&](ModuleOp op) -> WalkResult {
```
We've checked to make sure that this correctly handles the root module in non-fusion cases?

yiqian1 replied:
Yeah, tuning Gemm has passed.

krzysz00 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -857,6 +866,18 @@ pipeline {
                             }
                         }
                     }
+                    //stage("Test Fusion") {
```
Why's this commented out?

yiqian1 replied:
Because we don't have fusion tuning DB yet, it will fail in nightly CI. I'll enable it after the weekly generates fusion tuning DBs.

krzysz00 replied:
Ah

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -41,10 +41,14 @@
 class MLIRPaths:
     rocmlir_gen_path: str
     rocmlir_driver_path: str
+    rocmlir_opt_path:str
```
Nit: spaces around the colons

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -372,16 +387,19 @@ def fromCommandLine(cls, argv, arch):
-                if layout is not None and layout != arg:
-                    raise ValueError(f"Mixed layouts: {layout} vs {arg}")
+                filterLayout = arg
+                #if layout is not None and layout != arg:
```
... that's a perfectly good error message, why's it being removed?

yiqian1 replied:
Some fusion test cases have mixed layouts. So I removed the constrain that all layouts must be the same. We can add it back after we fix the mixed layouts if we want.

krzysz00 replied:
Ok, yeah, dropping the same-layout constraint is a good call, and if we're doing that we might as well actually remove it

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +822,127 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+                return parts[0].strip()
+                # Stop processing lines after finding the first matching line
+    # Not a valid test
+    print("WARNING: cannot find RUN command in the test file")
```
Give the name of the test file here

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +822,127 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+
+    if ("-migraphx-to-tosa" in firstCommand):
+        rocmlirOptCommand = [paths.mlir_paths.rocmlir_opt_path, '-migraphx-to-tosa', file_path]
+        rocmlirDriverCommand = [paths.mlir_paths.rocmlir_driver_path, '-host-pipeline', 'partition,highlevel', '-targets', getChip()]
```
Strong opinion: Add `,runner` to the host pipeline, and then, instead of using `xmir-runner`, use `mlir-cpu-runner`. This makes absolutely sure we don't benchmark GPU compilations or the like.

yiqian1 replied:
I tried to use mlir-cpu-runner. For some unknown reason, rocprof didn't generate result.stats.csv without giving any error messages.

krzysz00 replied:
Did you pass in `runner` as part of the `rocmlir-driver` host pipeline? Also, if you printed out the commands and ran them manually, did you get reasonable results?

yiqian1 replied:
Yes, I did pass in runner and run the commands manually. The tests were successful. But rocprof didn't generate the file result.stats.csv.

krzysz00 replied:
If you use the runner pipeline, you should be able to use `mlir-cpu-runner -O2` like all the other perf tests

krzysz00 replied:
(I'll note that `,runner` goes on the `rocmlir-driver` invocation after `rocmlir-opt`

yiqian1 replied:
For some reason that I don't understand, rocprof doesn't generate results.stats.csv if I use cpu-runner here.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +822,127 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+            # find the best perf_config
+            bestPerf =""
+            if tuningDb:
+                if (arch, filename, testVector) in tuningDb:
```
Reject putting file names in the tuning DB.

The tuning DB should, for now, just key off the gemm or convolution, ignoring the fusions.

yiqian1 replied:
Some test cases have the same tuning keys and the tuning-driver gave different best configs. Are they just noise? Is it ok to choose any best config?

krzysz00 replied:
Ok, so
1. How distinct were those perf numbers?
2. Which of those configs, if any, corresponded to the config for just that GEMM without any fusions?

krzysz00 replied:
That is to say, we should investigate that further. You'll also want to make sure you've merged in the tuner fixes from #1030 and #1032

yiqian1 replied:
1. The perf numbers are not distinct much. For one test case I've looked at, there is just one parameter off by 4. But I didn't check other test cases.
2. Don't know yet.
I haven't merged #1030 and #1032 though.

yiqian1 replied:
> Ok, so
> 
> 1. How distinct were those perf numbers?
> 2. Which of those configs, if any, corresponded to the config for just that GEMM without any fusions?

None of them did.

yiqian1 replied:
> 
> 1. How distinct were those perf numbers?
 The TFlops resulting from different perf configs were very close, with differences of less than 1%. I think that was simply normal variation between runs .

yiqian1 replied:
> That is to say, we should investigate that further. You'll also want to make sure you've merged in the tuner fixes from #1030 and #1032

Nothing changed with PR#1030 and #1032.

krzysz00 reviewed (changes requested):
1) I'm seeing a lot of copy-paste that could be refactored
2) I do not, at all, want the file name in the tuning database. The tuning database should key off the gemm/convolution portion of the test.
3) There's a bunch of stuff commented out that shouldn't be  - if you're going to remove it, remove it and explain why.

jerryyin committed:
Amending quantized ops type requirement according to migraphx (#1031)

manupak committed:
Add migraphx.mul to illegal ops (#1028)

Currently the dialect transformation does not
define mul as an illegal op. Hence the tablegen
define lowering is not taking place. This commit
fixes that and tests to assure that.

giuseros committed:
Fixing benchmark/tuner issues (#1030)

* Fixing benchmark/tuner issues

* Bug fixing

* Bug fixing - 2

* Addressing review feedbacks - 2

---------

Co-authored-by: Giuseppe Rossini <grossini@amd.com>

giuseros committed:
Fix a typo in PR #1030 (#1032)

* Fix a typo in PR #1030

* Fix an allocation bug

---------

Co-authored-by: Giuseppe Rossini <grossini@amd.com>

yiqian1 committed:
Remove filename from tuning DB; Refactor

yiqian1 committed:
Remove filename from tuning DB; Refactor

yiqian1 committed:
Merge branch 'fusion-e2e' of github.com:ROCmSoftwarePlatform/rocMLIR into fusion-e2e

yiqian1 committed:
Add fusion reports

yiqian1 committed:
Extract configs from fusion tests

yiqian1 committed:
Remove an unneeded verify-mode

yiqian1 committed:
Fix a bug that incorrectly detected a gemm operation

yiqian1 committed:
Typo

krzysz00 commented on mlir/include/mlir/Dialect/MIGraphX/MIGraphXOps.td:
```diff
@@ -160,30 +160,32 @@ def MIGraphX_FloorOp :
 def MIGraphX_QuantizeLinearOp :
     MIGraphX_Op<"quantizelinear">,
-    Arguments<(ins TensorOf<[F32]>:$input,
+    Arguments<(ins TensorOf<[F32, F16]>:$input,
```
What's this about?

krzysz00 commented on mlir/include/mlir/Dialect/MIGraphX/MIGraphXOps.td:
```diff
@@ -160,30 +160,32 @@ def MIGraphX_FloorOp :
-                   Optional<TensorOf<[I32]>>:$bias)>,
-	  Results<(outs TensorOf<[I8]>:$output)> {
+                   Optional<TensorOf<[AnyInteger]>>:$bias)>,
+	  Results<(outs TensorOf<[AnyInteger]>:$output)> {
```
I'll note that the API here needs to exactly match MIGraphX's IR

krzysz00 replied:
... wait, hold on, are these changes actually part of your PR or is the diff just noisy?

krzysz00 commented on mlir/utils/performance/createFusionPerformanceReports.py:
```diff
@@ -0,0 +1,50 @@
+        means.loc["All"] = df[COLUMNS_TO_AVERAGE].agg(reportUtils.geoMean)
+        means.to_csv(chip + '_' + op + '_' + reportUtils.PERF_STATS_REPORT_FUSION_FILE)
+
+    toHighlight = [""]
```
This should be `[]` instead

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -372,16 +384,19 @@ def fromCommandLine(cls, argv, arch):
-                if layout is not None and layout != arg:
-                    raise ValueError(f"Mixed layouts: {layout} vs {arg}")
+                inputLayout = arg
+                #if layout is not None and layout != arg:
```
Please remove instead of commenting out.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -414,28 +429,39 @@ def fromCommandLine(cls, argv, arch):
                    group, arch)
 
-    def __init__(self, dtype: str, direction: str, layout: str,
+    def __init__(self, dtype: str, direction: str, fLayout: str, iLayout:str, oLayout:str,
```
Let's go with `filterLayout`, `inputLayout`, and `outputLayout`

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -414,28 +429,39 @@ def fromCommandLine(cls, argv, arch):
                 continue
 
-        return cls(dataType, direction, layout, n, c, hi, wi, k, y, x,
+        if layout is None:
```
More substantial comment: let's remove the `layout` variable all together, since we're now tracking filter, input, and output layouts seperately.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -414,28 +429,39 @@ def fromCommandLine(cls, argv, arch):
+        #self.inputLayout = layout.lower()
+        #self.outputLayout = self.MLIR_OUTPUT_LAYOUTS[layout]
+        self.filterLayout = ''.join(self.filterLayoutMap.get(c).lower() for c in fLayout)
+        self.inputLayout = ''.join(self.inputLayoutMap.get(c).lower() for c in iLayout)
```
Given the context, this should probably be `.get(c, c)` so that MLIR-style layouts (like `ngkhw`) don't throw exceptions.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +819,133 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
 
+def findRunCommand(filename, includes=['RUN']):
+    with open(filename, 'r') as f:
+        for line in f:
```
Instead of paramtrizing on `includes`, maybe this should just be a regex?

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +819,133 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+                # Stop processing lines after finding the first matching line
+    # Not a valid test
+    print("WARNING: cannot find valid RUN command in ", filename)
+    return ""
```
`None`, perhaps?

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +819,133 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+    print("WARNING: cannot find valid RUN command in ", filename)
+    return ""
+
+def getTestVector(file_path, paths:Paths):
```
`paths: Paths`

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +819,133 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+
+def getTestVector(file_path, paths:Paths):
+    firstCommand = findRunCommand(file_path)
+    if (not firstCommand):
```
Redundant parentheses. Also, maybe `firstCommand is None` given above suggestions.

krzysz00 commented on mlir/utils/performance/perfRunner.py:
```diff
@@ -793,6 +819,133 @@ def benchmarkMIOpenWithMLIRKernels(configs, arch, filename, paths: Paths):
+    return result[1]
+
+def generateProfilingProcesses(file_path, rocmlirGenArgs, paths: Paths):
+    firstCommand = findRunCommand(file_path, ['RUN', 'xmir-runner'])
```
Yeah, this might be better parametrized by `hasRunner` as a boolean option and then filtering on the regex output.

krzysz00 commented on mlir/utils/performance/tuningRunner.py:
```diff
@@ -134,6 +143,43 @@ def tuneMLIRKernels(configs, confClass, paths: Paths, options: Options):
+    allConfigs = []
+    opType=Operation.FUSION
+    for filename in os.listdir(test_dir):
+        if filename.endswith(".mlir"):
```
This is `glob`

krzysz00 reviewed (commented):
1. Your PR's in a weird state where it's reverting a whole bunch of changes made on `develop`. Could you please fix that?
2. Various python nitpicks
3. Let's try not using `xmir-runner` here - I'd like the program we benchmark to be very minimal, and `xmir-runner` does a lot of compilation in the "runner".

yiqian1 commented:
It's hard to fix the mess caused by a bad merge. I'll start a new PR.
